### ISSUE TITLE
WARNING:tensorflow:Gradients do not exist for variables

### ISSUE THREAD

**Issue by talhaanwarch** [tensorflow/tensorflow#37501]
While training BERT on TPU i am getting these warnings and my precision and recall is zero while accuracy is 100

```
Train for 45205 steps, validate for 206 steps
WARNING:tensorflow:Gradients do not exist for variables ['tf_bert_model_1/bert/pooler/dense/kernel:0', 'tf_bert_model_1/bert/pooler/dense/bias:0'] when minimizing the loss.
WARNING:tensorflow:Gradients do not exist for variables ['tf_bert_model_1/bert/pooler/dense/kernel:0', 'tf_bert_model_1/bert/pooler/dense/bias:0'] when minimizing the loss.
WARNING:tensorflow:Gradients do not exist for variables ['tf_bert_model_1/bert/pooler/dense/kernel:0', 'tf_bert_model_1/bert/pooler/dense/bias:0'] when minimizing the loss.
WARNING:tensorflow:Gradients do not exist for variables ['tf_bert_model_1/bert/pooler/dense/kernel:0', 'tf_bert_model_1/bert/pooler/dense/bias:0'] when minimizing the loss.

```

Here is google colab file i am using
https://colab.research.google.com/drive/1l0Eoram6vJRK5xQBmnCx-eU1e46higlo

updated colab file
https://colab.research.google.com/drive/1D-eKgddRHROG39R2iyqE3qR7w0N0OHAB

**Comment by amahendrakar** [comment]
@talhaanwarch,
I was unable to reproduce the issue as the file `task_a_distant.tsv` is missing. Please find the gist of it [here](https://colab.sandbox.google.com/gist/amahendrakar/0768ebd3095d4a8410ba8407f45e595e/37501.ipynb).
Could you please share all the supporting files required to run the given code? Thanks!

**Comment by talhaanwarch** [comment]
@amahendrakar  please check this file, i have updated.
https://colab.research.google.com/drive/1D-eKgddRHROG39R2iyqE3qR7w0N0OHAB

**Comment by amahendrakar** [comment]
@talhaanwarch,
On running the updated colab gist, I got the metrics `binary_accuracy: 0.6635 - precision: 0.3663 - recall: 0.0120`. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/5c2225faee2306e9fce0550bfc55aec1/37501-v2.ipynb). Thanks!

**Comment by talhaanwarch** [comment]
but the warning is still there. `WARNING:tensorflow:Gradients do not exist for variables ['tf_bert_model/bert/pooler/dense/kernel:0', 'tf_bert_model/bert/pooler/dense/bias:0'] when minimizing the loss.`  I am thinking it should be an error instead of warning. becuase if gradient donot exist, how it will be updated?

**Comment by gowthamkpr** [comment]
@talhaanwarch Please take a look at this [issue](https://github.com/huggingface/transformers/issues/2256) which is very similar to this issue.

**Comment by gowthamkpr** [comment]
@talhaanwarch Can yo please respond to the above comment so that we can take the discussion further. Thanks!

**Comment by talhaanwarch** [comment]
i requested at [issue ](https://github.com/huggingface/transformers/issues/2256) if he can update his colab notebook with solution, but for me i am unable to figure out the solution

**Comment by saahiluppal** [comment]
I'm getting same error. No solution found so far.

**Comment by LunarBerial** [comment]
same error founded.

**Comment by nguyenphuhien13** [comment]
same error for me

**Comment by gowthamkpr** [comment]
@talhaanwarch Please post your issue [here](https://github.com/google-research/bert/issues). Is this happening only on TPUs ? or is it happening on GPUs too?

**Comment by acse-ncv19** [comment]
I have a similar warning message using TF2.2 on Colab GPU:
WARNING:tensorflow:Gradients do not exist for variables ['Variable:0'] when minimizing the loss.

**Comment by talhaanwarch** [comment]
@gowthamkpr  i am getting this error on TPU only

**Comment by gowthamkpr** [comment]
@talhaanwarch The warning occurs on GPU too. Its not specific to TPU. Please find the the gist [here](https://colab.research.google.com/gist/gowthamkpr/88af1c86ace8bdcaf82e5c7228d83b31/37501-v2.ipynb). As mentioned in this [issue comment](https://github.com/huggingface/transformers/issues/1727#issuecomment-568058704), this is not an bug. It means that these variables are not updated during training.

If you have more questions, please post the issue [here](https://github.com/google-research/bert/issues) or in stackoverflow. Thanks!

**Comment by gajeshladhar** [comment]
I'm also getting this warning ! my embedding layers and attention layers are not updating and remains fixed as initialized. if anyone have solution please tell me.

**Comment by gajeshladhar** [comment]
model architecture 
`# Defining all Layers...

embd=Embedding(input_dim=len(vocab),output_dim=100,name="embd")

lstm1=Bidirectional(LSTM(units=100,return_sequences=True,name="lstm1"),name="bd1")
lstm2=Bidirectional(LSTM(units=100,return_sequences=True,name="lstm2"),name="bd2")

# Attention Layers to make seq-length=21 
# Developed by Gajesh Ladhar
attention_layer=Attention_Model(21,200)


dense1=Dense(units=80,name="dense1",kernel_regularizer="l2")
dropout1=Dropout(0.5)
act1=Activation('sigmoid')

dense2=Dense(units=50,name="dense2",kernel_regularizer="l2")
dropout2=Dropout(0.4)
act2=Activation('sigmoid')

dense3=Dense(units=30,name="dense3",kernel_regularizer="l2")
dropout3=Dropout(0.3)
act3=Activation('sigmoid')

dense4=Dense(units=len(classes),name="dense4")
dropout4=Dropout(0.2)
output=Activation('softmax')
`
Forward pass 

`def forward_pass(X):
  t=embd(X)
 
  t=lstm1(t)
  
  t=lstm2(t)
  

  # Input to Attention Layer
  t=attention_layer(t)
  
  
  t=dense1(t)
  t=dropout1(t)
  t=act1(t)

  t=dense2(t)
  t=dropout2(t)
  t=act2(t)

  t=dense3(t)
  t=dropout3(t)
  t=act3(t)
  
  t=dense4(t)
  t=dropout4(t)
  t=output(t)

  return t`

Attention model :

`class Attention_Model():
  def __init__(self,seq_length,units):
    self.seq_length=seq_length
    self.units=units
    self.lstm=LSTM(units=units,return_sequences=True,return_state=True)
    

  def get_lstm_s(self,seq_no):
    input_lstm=tf.expand_dims(tf.reduce_sum(self.X*(self.alphas[:,:,seq_no:seq_no+1]),axis=1),axis=1)
    a,b,c=self.lstm(input_lstm)
    self.output[:,seq_no,:]=a[:,0,:]

    return b

  def __call__(self,X):
    self.X=X
    self.output=np.zeros(shape=(self.X.shape[0],self.seq_length,self.units))
    self.dense=Dense(units=self.seq_length)
    self.softmax=Softmax(axis=1)
    

    for i in range(self.seq_length+1):
      if i==0 :
        s=np.zeros(shape=(self.X.shape[0],self.units))
      else :
        s=self.get_lstm_s(i-1)
      if(i==self.seq_length):
        break 
      
      s=RepeatVector(self.X.shape[1])(s)
      concate_X=np.concatenate([self.X,s],axis=-1)
      
      self.alphas=self.softmax(self.dense(concate_X))

    return self.output
      
`

is anything wrong with implementation or something else ?

**Comment by SmileTM** [comment]
@amahendrakar  
how to remove these WARNING from output Info ???

**Comment by amahendrakar** [comment]
@SmileTM,
To disable the warning logs, try changing the log level using the below code.
```
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2' 
import tensorflow as tf
```
If you need further help, please submit a new issue from [this link](https://github.com/tensorflow/tensorflow/issues/new/choose) and fill in the template, so that we can track the issue there. Thanks!

**Comment by zccyman** [comment]
https://stackoverflow.com/questions/57144586/tensorflow-gradienttape-gradients-does-not-exist-for-variables-intermittently
````
optimizer.apply_gradients(
    (grad, var) 
    for (grad, var) in zip(gradients, model.trainable_variables) 
    if grad is not None
)
````

**Comment by FathiMahdi** [comment]
For me, I just reduce the batch size to 5 and then ignored the waring
